src/dotty/tools/dotc/reporting/diagnostic/messages.scala

```diff
@@ -318,4 +319,91 @@ object messages {
         |$code2
         |""".stripMargin
   }
+
+  val implicitClassRestrictionsText =
+    """For a full list of restrictions on implicit classes visit
+      |http://docs.scala-lang.org/overviews/core/implicit-classes.html""".stripMargin
+
+  case class TopLevelImplicitClass(cdef: untpd.TypeDef)(implicit ctx: Context)
+    extends Message(10) {
+    val kind = "Syntax"
+
+    val msg = hl"""|An ${"implicit class"} may not be top-level"""
+
+    val explanation = {
+      val TypeDef(name, impl @ Template(constr0, parents, self, _)) = cdef
+      val exampleArgs = constr0.vparamss(0).map(_.withMods(Modifiers()).show).mkString(", ")
+      val exampleBodyDefs = impl.body.collect {
+        case t if !t.isEmpty => t.show
+      }
+      val exampleBody = if (exampleBodyDefs.isEmpty) ""
+      else exampleBodyDefs.mkString(
+        start = " {\n    ",
+        sep = "\n\n    ",
+        end = "\n  }"
+      )
+      hl"""|There may not be any method, member or object in scope with the same name as the
+           |implicit class and a case class automatically gets a companion object with the same name
+           |created by the compiler which would cause a naming conflict if it were allowed.
+           |
+           |$implicitClassRestrictionsText
+           |
+           |To resolve the conflict declare ${cdef.name} inside of an ${"object"} then import the class
+           |from the object at the use site if needed, for example:
+           |
+           |object Implicits {
+           |  implicit class ${cdef.name}($exampleArgs)$exampleBody
+           |}
+           |
+           |import Implicits.${cdef.name}""".stripMargin
+    }
+  }
```

Comment:
I think it would be nice to make it more explicit that the `import Implicits.Foo` is a at the use site, for example by writing:
```scala
object Implicits {
  implicit class Foo(val bar: Int, val baz: String){
 ...
 }
}

...

// At the use site:
import Implicits.Foo
```